### PR TITLE
[lex.ccon,lex.string] Remove redundant repetition of the grammar.

### LINKED ISSUE #3636: [lex] Replace "character literal" with references to grammar term character-literal
From CWG discussion of P2029R0.

Remove [lex.ccon] paragraph 1, which defines "character literal" as exactly the same thing as a (grammar term) *character-literal*, and replace all references to "character literal" with references to the grammar term *character-literal*.

Also remove [lex.string] paragraph 1, which defines in plain English the grammar term *string-literal* to be the same thing it already is in the grammar.

### PR DESCRIPTION
**Pull request by jensmaurer**
Also use the grammar non-terminals character-literal and
string-literal throughout the standard.

Fixes #3636.

### CODE REVIEW

File: source/compatibility.tex [hunk 1]
```diff
@@ -97,7 +97,7 @@
 \end{codeblock}
 
 \howwide
-Programs that have a legitimate reason to treat string literals
+Programs that have a legitimate reason to treat \grammarterm{string-literal}s
```
Comment (zygoloid) [inline source/compatibility.tex]: I think this means "string literal objects" not *string-literal*s.

File: source/compatibility.tex [hunk 2]
```diff
@@ -64,7 +64,7 @@
 \diffref{lex.string}
 \change
 String literals made const.\\
-The type of a string literal is changed
+The type of a \grammarterm{string-literal} is changed
```
Comment (zygoloid) [inline source/compatibility.tex]: (Digression: I think we are presenting *string-literal*s very strangely. We should have something like:

> *string-literal-token*:
> &nbsp; &nbsp; \<current *string-literal* productions>

> *string-literal*:
> &nbsp; &nbsp; *string-literal-token*
> &nbsp; &nbsp; *string-literal* *string-literal-token*

And we should delete phase 6 of translation and move the string literal concatenation and string literal objects wording out of [lex] into [expr]. Then we could clearly distinguish between the cases where we're talking about *string-literal*s, such as this, and the cases where we're talking about *string-literal-token*s, which covers a bunch of the other changes in this PR.)

File: source/lex.tex [hunk 3]
```diff
@@ -319,8 +319,8 @@
 phases 3 through 6. The categories of preprocessing token are: header names,
 placeholder tokens produced by preprocessing \tcode{import} and \tcode{module} directives
 (\grammarterm{import-keyword}, \grammarterm{module-keyword}, and \grammarterm{export-keyword}),
-identifiers, preprocessing numbers, character literals (including user-defined character
-literals), string literals (including user-defined string literals), preprocessing
+identifiers, preprocessing numbers, \grammarterm{character-literal}s (including user-defined character
+literals), \grammarterm{string-literal}s (including user-defined string literals), preprocessing
```
Comment (zygoloid) [inline source/lex.tex]: This change is incorrect. User-defined {character,string} literals are not *character-literal* or *string-literal* tokens.

File: source/lex.tex [hunk 4]
```diff
@@ -334,8 +334,8 @@
 circumstances during translation phase 4, white space (or the absence
 thereof) serves as more than preprocessing token separation. White space
 can appear within a preprocessing token only as part of a header name or
-between the quotation characters in a character literal or string
-literal.
+between the quotation characters in a \grammarterm{character-literal} or
+\grammarterm{string-literal}.
```
Comment (zygoloid) [inline source/lex.tex]: This is also unclear in the presence of UDLs.

File: source/lex.tex [hunk 5]
```diff
@@ -256,8 +256,8 @@
 If a \grammarterm{universal-character-name} outside
 the \grammarterm{c-char-sequence}, \grammarterm{s-char-sequence}, or
 \grammarterm{r-char-sequence} of
-a character or
-string literal corresponds to a control character or
+a \grammarterm{character-literal} or
+\grammarterm{string-literal} corresponds to a control character or
```
Comment (zygoloid) [inline source/lex.tex]: This change is unclear in the presence of UDLs. (*string-literal* is both a syntactic production and a morphological production, and it's unclear what this means.)

File: source/expressions.tex [hunk 6]
```diff
@@ -1158,7 +1158,7 @@
 \grammarterm{literal}
 is a primary expression.
 Its type depends on its form\iref{lex.literal}.
-A string literal is an lvalue; all other literals are prvalues.
+A \grammarterm{string-literal} is an lvalue; all other literals are prvalues.
```
Comment (zygoloid) [inline source/expressions.tex]: Hmm. The existing wording here is just plain wrong. UDLs can have any value category.

File: source/compatibility.tex (hunk 1 above)
Comment (jensmaurer) [reply source/compatibility.tex]: "string literal objects" it is.

File: source/compatibility.tex (hunk 2 above)
Comment (jensmaurer) [reply source/compatibility.tex]: Uh, is introducing new grammar-terms and deleting a phase of translation considered editorial these days?

File: source/lex.tex (hunk 3 above)
Comment (jensmaurer) [reply source/lex.tex]: Reverted.

File: source/lex.tex (hunk 4 above)
Comment (jensmaurer) [reply source/lex.tex]: Reverted.

File: source/lex.tex (hunk 5 above)
Comment (jensmaurer) [reply source/lex.tex]: We are in [lex], so grammar is assumed to refer to the lexical grammar. And UDLs do use the productions string-literal and character-literal.

File: source/expressions.tex (hunk 6 above)
Comment (jensmaurer) [reply source/expressions.tex]: Applied a larger hammer.

File: source/compatibility.tex (hunk 2 above)
Comment (zygoloid) [reply source/compatibility.tex]: No, hence the "digression" part. :) I think that is something we should take to CWG for the C++23 cycle.

File: source/expressions.tex [hunk 7]
```diff
@@ -1152,13 +1152,16 @@
...
-Its type depends on its form\iref{lex.literal}.
-A string literal is an lvalue; all other literals are prvalues.
+A \grammarterm{literal} is a primary expression.
+A \grammarterm{user-defined-literal} is interpreted as a function call
+as described in \ref{lex.ext};
+the type and value category of such a literal is determined by the
+rules for function calls\iref{expr.call}.
+For other literals,
+the type is determined by its form as specified in \ref{lex.literal};
+a \grammarterm{string-literal} is an lvalue; all other literals are prvalues.
```
Comment (zygoloid) [inline source/expressions.tex]: This is a bit too semicolon-heavy for my tastes. How about something like:

> A *literal* is a primary expression, whose type is determined based on its form as specified in [lex.literal]. A *string-literal* is an lvalue, a *user-defined-literal* has the same value category as the corresponding operator call expression described in [lex.ext], and any other *literal* is a prvalue.

File: source/lex.tex [hunk 8]
```diff
@@ -127,16 +127,16 @@
...
+\grammarterm{string-literal}, as well as each escape sequence and \grammarterm{universal-character-name} in a
+\grammarterm{character-literal} or a non-raw string literal, is converted to the corresponding
 member of the execution character set~(\ref{lex.ccon}, \ref{lex.string}); if
 there is no corresponding member, it is converted to an \impldef{converting
 characters from source character set to execution character set} member other
 than the null (wide) character.\footnote{An implementation need not convert all
 non-corresponding source characters to the same execution character.}
 
-\item Adjacent string literal tokens are concatenated.
+\item Adjacent \grammarterm{string-literal} tokens are concatenated.
```
Comment (zygoloid) [inline source/lex.tex]: This is not correct; as specified in [lex.ext]p8, this also includes *user-defined-string-literal* tokens, and we're definitely talking about complete tokens and not lexer rules here.

File: source/lex.tex (hunk 5 above)
Comment (zygoloid) [reply source/lex.tex]: I don't think this is clear. (Similar questions: do the reservations in [lex.name]/3 apply to *ud-suffix*es, or only to identifier tokens? Is "orbit" as an identifier ill-formed because it contains the identifier "or" which is reserved and "shall not be used otherwise"?)

Adding an "(including within a *user-defined-literal*)" or similar would make this completely clear. If you think this is already clear normatively, maybe use a footnote "This includes a *string-literal* or *character-literal* within a *user-defined-string-literal* or *user-defined-character-literal* token." rather than a parenthetical?

File: source/templates.tex [hunk 9]
```diff
@@ -1221,7 +1221,7 @@
 or be the address of (respectively):
 \begin{itemize}
 \item a temporary object\iref{class.temporary},
-\item a string literal\iref{lex.string},
+\item a \grammarterm{string-literal}\iref{lex.string},
```
Comment (zygoloid) [inline source/templates.tex]: string literal object

File: source/lex.tex (hunk 5 above)
Comment (jensmaurer) [reply source/lex.tex]: I don't think that [lex] is clear normatively, at all, in particular the little lexer sub-grammar used here with unclear relation to the "big" grammar.

File: source/templates.tex [hunk 10]
```diff
@@ -1269,8 +1269,8 @@
   @\commentellip@
 };
 
-X<const char*, "Studebaker"> x; // error: string literal as \grammarterm{template-argument}
-X<const char*, "Knope" + 1> x2; // error: subobject of string literal as \grammarterm{template-argument}
+X<const char*, "Studebaker"> x; // error: \grammarterm{string-literal} as \grammarterm{template-argument}
+X<const char*, "Knope" + 1> x2; // error: subobject of \grammarterm{string-literal} as \grammarterm{template-argument}
```
Comment (zygoloid) [inline source/templates.tex]: "string literal object" (x2)

File: source/templates.tex (hunk 9 above)
Comment (jensmaurer) [reply source/templates.tex]: Fixed.

File: source/templates.tex (hunk 10 above)
Comment (jensmaurer) [reply source/templates.tex]: Fixed.

File: source/lex.tex (hunk 8 above)
Comment (jensmaurer) [reply source/lex.tex]: Reverted.

File: source/lex.tex (hunk 5 above)
Comment (jensmaurer) [reply source/lex.tex]: Fixed.

File: source/expressions.tex (hunk 7 above)
Comment (jensmaurer) [reply source/expressions.tex]: Fixed.
